pkg/sfu/buffer/rtpstats_receiver.go

```diff
@@ -326,26 +351,83 @@ func (r *RTPStatsReceiver) SetRtcpSenderReportData(srData *RTCPSenderReportData)
 		}
 	}
 
-	r.srNewest = &srDataCopy
+	ntpTime := srDataCopy.NTPTimestamp.Time()
+	propagationDelay := srDataCopy.At.Sub(ntpTime)
 	if r.srFirst == nil {
 		r.srFirst = &srDataCopy
+		r.propagationDelay = propagationDelay
+		r.logger.Debugw(
+			"intializing propagation delay",
+			"propagationDelay", r.propagationDelay,
+			"first", r.srFirst,
+			"last", r.srNewest,
+			"current", &srDataCopy,
+		)
+	} else {
+		delta := propagationDelay - r.propagationDelay
+		if delta.Abs() > r.propagationDelay.Abs()*time.Duration(cPropagationDelayChangeThresholdMaxFactor) {
+			r.logger.Infow(
+				"sharp increase in propagation delay, skipping",
+				"propagationDelay", r.propagationDelay,
+				"receivedPropagationDelay", propagationDelay,
+				"first", r.srFirst,
+				"last", r.srNewest,
+				"current", &srDataCopy,
+			)
+			r.propagationDelayChangeHighCount++
+			if r.propagationDelayChangeHighStartTime.IsZero() {
+				r.propagationDelayChangeHighStartTime = time.Now()
+			}
+
+			if r.propagationDelayChangeHighCount >= cPropagationDelayChangeHighResetNumReports && time.Since(r.propagationDelayChangeHighStartTime) >= cPropagationDelayChangeHighResetWait {
+				r.logger.Infow(
+					"re-intializing propagation delay",
+					"propagationDelay", r.propagationDelay,
+					"newPropagationDelay", propagationDelay,
+					"first", r.srFirst,
+					"last", r.srNewest,
+					"current", &srDataCopy,
+				)
+				r.propagationDelay = propagationDelay
+				r.propagationDelayChangeHighCount = 0
+				r.propagationDelayChangeHighStartTime = time.Time{}
+			}
+		} else {
+			r.propagationDelayChangeHighCount = 0
+			r.propagationDelayChangeHighStartTime = time.Time{}
+			if delta.Abs() > cPropagationDelayChangeThresholdMin {
+				factor := cPropagationDelayFallFactor
+				if propagationDelay > r.propagationDelay {
+					factor = cPropagationDelayRiseFactor
+				}
+				r.logger.Debugw(
+					"adapting propagation delay",
+					"propagationDelay", r.propagationDelay,
+					"receivedPropagationDelay", propagationDelay,
+					"newPropagationDelay", r.propagationDelay+time.Duration(factor*float64(propagationDelay-r.propagationDelay)),
+					"first", r.srFirst,
+					"last", r.srNewest,
+					"current", &srDataCopy,
+				) // TODO-REMOVE
```

Comment:
This will be noisy. Will delete after some collecting some data to check how much this changes.